### PR TITLE
Move CRQ reset to process context

### PR DESCRIPTION
**Pull request by ghost**
H_FREE_CRQ calls may return long busy return codes. Best practice
is to sleep through the delay, which cannot be performed in atomic
context.  This change uses msleep instead of mdelay to handle long
busy codes, and re-organizes CRQ reset logic to ensure
H_FREE_CRQ calls always occur in process context.

### CODE REVIEW

File: ibmvmc/ibmvmc.c [hunk 1]
```diff
@@ -95,10 +96,13 @@ static inline void h_free_crq(uint32_t unit_address)
 	long rc = 0;
 
 	do {
-		if (H_IS_LONG_BUSY(rc))
-			mdelay(get_longbusy_msecs(rc));
+		if (H_IS_LONG_BUSY(rc)) {
+			pr_debug("ibmvmc: h_free_crq: sleeping for %u msec\n", get_longbusy_msecs(rc));
```
Comment (seroyer) [inline ibmvmc/ibmvmc.c]: I don't think we need to leave this pr_debug in from development.

File: ibmvmc/ibmvmc.c [hunk 2]
```diff
@@ -95,10 +96,13 @@ static inline void h_free_crq(uint32_t unit_address)
...
+			msleep(get_longbusy_msecs(rc));
+		}
 
 		rc = plpar_hcall_norets(H_FREE_CRQ, unit_address);
+		pr_debug("ibmvmc: h_free_crq: rc = %lx\n", rc);
```
Comment (seroyer) [inline ibmvmc/ibmvmc.c]: I don't think we need to leave this pr_debug in from development.

File: ibmvmc/ibmvmc.c [hunk 3]
```diff
@@ -655,12 +662,13 @@ static int ibmvmc_close(struct inode *inode, struct file *file)
 			return -EIO;
 		}
 
-		/* TODO locking? */
+		spin_lock_irqsave(&hmc->lock, flags);
```
Comment (seroyer) [inline ibmvmc/ibmvmc.c:664]: Was this change needed for some reason for the sleep in process context?  Or you just cleaned this up because of the TODO?

Comment (ghost) [reply ibmvmc/ibmvmc.c:664]: Partially because of the TODO, but the intention was to prevent sending close CRQs for the HMC if the HMC was already returned by the reset task. (We also do a state check and CRQ send under the HMC lock in ibmvmc_write)

File: ibmvmc/ibmvmc.c (hunk 2 above)
Comment (ghost) [reply ibmvmc/ibmvmc.c]: I'll remove these.

**Comment by seroyer** [comment]
This looks good, but please squash into a single commit.

File: ibmvmc/ibmvmc.c [hunk 4]
```diff
@@ -156,6 +157,8 @@ static void ibmvmc_release_crq_queue(struct crq_server_adapter *adapter)
 	free_irq(vdev->irq, (void *)adapter);
 	tasklet_kill(&adapter->work_task);
```
Comment (bgly) [inline ibmvmc/ibmvmc.c:159]: if (adapter->reset_task)

File: ibmvmc/ibmvmc.c [hunk 5]
```diff
@@ -1743,6 +1808,10 @@ static int ibmvmc_probe(struct vio_dev *vdev, const struct vio_device_id *id)
 	dev_set_drvdata(&vdev->dev, adapter);
 
 	return 0;
+
+crq_failed:
+	kthread_stop(adapter->reset_task);
```
Comment (bgly) [inline ibmvmc/ibmvmc.c:1817]: adapter->reset_task = NULL;

File: ibmvmc/ibmvmc.c [hunk 6]
```diff
@@ -1723,11 +1780,19 @@ static int ibmvmc_probe(struct vio_dev *vdev, const struct vio_device_id *id)
 	pr_debug("ibmvmc: Probe: liobn 0x%x, riobn 0x%x\n", adapter->liobn,
 			adapter->riobn);
 
+	init_waitqueue_head(&adapter->reset_wait_queue);
+	adapter->reset_task = kthread_run(ibmvmc_reset_task, adapter, "ibmvmc");
+	if (IS_ERR(adapter->reset_task)) {
+		pr_err("ibmvmc: Failed to start reset thread\n");
+		ibmvmc.state = ibmvmc_state_failed;
```
Comment (bgly) [inline ibmvmc/ibmvmc.c:1789]: adapter->reset_task = NULL;

File: ibmvmc/ibmvmc.c [hunk 7]
```diff
@@ -1723,11 +1780,19 @@ static int ibmvmc_probe(struct vio_dev *vdev, const struct vio_device_id *id)
...
+		return -1;
```
Comment (bgly) [inline ibmvmc/ibmvmc.c]: might as well do return PTR_ERR(adapter->reset_task) instead of return -1

File: ibmvmc/ibmvmc.c [hunk 8]
```diff
@@ -1743,6 +1808,10 @@ static int ibmvmc_probe(struct vio_dev *vdev, const struct vio_device_id *id)
 	dev_set_drvdata(&vdev->dev, adapter);
 
 	return 0;
+
+crq_failed:
+	kthread_stop(adapter->reset_task);
+	return -1;
```
Comment (bgly) [inline ibmvmc/ibmvmc.c]: return PTR_ERR(adapter->reset_task) instead of return -1

Comment (ghost) [reply ibmvmc/ibmvmc.c]: What's the reasoning behind this? If we reach this point, reset_task was started correctly and wasn't the cause of failure.

File: ibmvmc/ibmvmc.c (hunk 7 above)
Comment (ghost) [reply ibmvmc/ibmvmc.c]: I can't do
`adapter->reset_task = NULL;`
`return PTR_ERR(adapter->reset_task);`
because that would return 0 (success)... I can do this instead:

`rc = PTR_ERR(adapter->reset_task);`
`adapter->reset_task = NULL;`
`return rc;`

File: ibmvmc/ibmvmc.c (hunk 8 above)
Comment (bgly) [reply ibmvmc/ibmvmc.c]: Sorry, just do the NULL from above here, you don't need to return the PTR_ERR. But change -1 to EPERM. I don't understand why we override the rc's in a lot of these cases.

File: ibmvmc/ibmvmc.c (hunk 7 above)
Comment (bgly) [reply ibmvmc/ibmvmc.c]: Yep, I just commented by line, so obviously use your 2nd version.

**Review by bgly — APPROVED** [review]

**Comment by seroyer** [comment]
Thanks!
